### PR TITLE
Customize scroll speed via props

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
FDT currently has no support for configuring the scroll speed through props.

## Description
Added a config prop that is used by FDT to customize props.
This prop contains all the values used for customizing scrolling.
The config selector will return the grouped configuration and also handle default/undefined values.
Most of these are used by ReactTouchHandler and ReactWheelHandler.

## Motivation and Context
Fixes (#220)
Helps (#309)

## How Has This Been Tested?
This isn't heavily tested, although I did verify that scroll behavior is changed according to values from config prop. I wanted to see if the current approach is in the right direction before I spend too much time on this.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/FixedDataTable.js [hunk 1]
```diff
@@ -463,52 +465,70 @@ class FixedDataTable extends React.Component {
...
     const { maxScrollX, scrollFlags, scrollX } = this.props;
     const { overflowX } = scrollFlags;
 
-    if (overflowX === 'hidden') {
+    if (overflowX === 'hidden' || delta == 0) {
       return false;
     }
 
-    delta = Math.round(delta);
-    if (delta === 0) {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: If we round delta and check it for zero, very small values will get ignored (most noticeable in touch scroll).

File: src/FixedDataTable.js [hunk 2]
```diff
@@ -463,52 +465,70 @@ class FixedDataTable extends React.Component {
...
+    if (typeof threshold === 'function' && threshold(delta)) {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: If threshold is a function, it determines whether scrolling is to be ignored or not.

File: src/FixedDataTable.js [hunk 3]
```diff
@@ -463,52 +465,70 @@ class FixedDataTable extends React.Component {
...
+      return false;
+    } else if (Math.abs(delta) <= threshold) {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: If threshold is a value, make sure magnitude of delta is greater than threshold for a scroll to occur.

File: src/ReactTouchHandler.js [hunk 4]
```diff
@@ -19,10 +19,13 @@
 import emptyFunction from 'emptyFunction';
 import requestAnimationFramePolyfill from 'requestAnimationFramePolyfill';
 
-var MOVE_AMPLITUDE = 1.6;
-var DECELERATION_AMPLITUDE = 1.6;
-var DECELERATION_FACTOR = 325;
-var TRACKER_TIMEOUT = 100;
+let DEFAULT_TOUCH_CONFIG = {
```
Comment (pradeepnschrodinger) [inline src/ReactTouchHandler.js]: Added new constant 'DECELERATION_THRESHOLD', which controls the autoscroll threshold.

File: src/Scrollbar.js [hunk 5]
```diff
@@ -503,11 +503,11 @@ class Scrollbar extends React.PureComponent {
   }
 
   getTouchX = (/*object*/ e) => {
-    return Math.round(e.targetTouches[0].pageX - e.target.getBoundingClientRect().x);
+    return Math.round(e.targetTouches[0].clientX - e.target.getBoundingClientRect().x);
```
Comment (pradeepnschrodinger) [inline src/Scrollbar.js]: So this fixes a small bug from #341, where clicking and dragging the scrollbar's divider is not in sync with the mouse when touch is used.

**Comment by pradeepnschrodinger** [comment]
## No breaking changes
I made sure that the default behavior is not changed, and is identical to calling it without the config prop.

## Sample usage:
```javascript
<Table
   ...
   config={{
      keyboardScrollAmplitude: 50,
      touchAmplitude: 0.5,
      touchDeceleration: 1,
      touchDecelerationFactor: .413,
      touchDecelerationThreshold: 0.3,
      touchThreshold: 0,
      touchTimeout: 100,
      wheelBaseFactor: 0.1,
      wheelPixelAmplitude: 1,
      wheelLineAmplitude: 1,
      wheelPageAmplitude: 1,
      wheelThreshold: 0,
   }}
   ...
>
```

## Changes in upstream files
Not sure if this is the right approach, but this makes a few changes in dom/ReactWheelHandler and dom/normalizeWheel, which are mostly just making use of the optional config argument. Also made some changes in ReactTouchHandler.

## Selector
The selector simply takes all props from config, and groups them under keyboard, wheel, and touch.
It also defaults values if undefined, and if any value is still undefined, removes it (so it won't get overridden in the lower levels)
There could be a lot of issues with variable names and how default values are used. 

## Is this the right approach?
@wcjordan Before I continue the work on this, I wanted to make sure I'm heading in the right direction. What are your thoughts on this?

**Comment by wcjordan** [comment]
I think this approach is good in general.  

I do have a concern with how many configs were exposing.  Generally I'd recommend exposing as few configs as possible to achieve the end-users needs.  Since you'll need to document every config, one exercise which can help with this is for every config try to write up a case or workflow where a user would need to change that behavior and ask yourself if it seems common.  If any config isn't used by a common workflow, I recommend hiding it until someone explicitly asks for it.

Wrt upstream, one hope is we'd eventually remove those from the repo and replace them with [fbjs](https://github.com/facebook/fbjs).  Could you take a look there and determine if fbjs can be used / extended for our purposes so we don't end up maintaining the code?  If not we should assume we'll need to fork and maintain it going forward.

**Comment by pradeepnschrodinger** [comment]
Closing this. Will reopen after I check out fbjs and see if I get time to do the changes.

**Pull request closed without merging**